packages/yew-agent/src/oneshot/provider.rs

```diff
@@ -0,0 +1,129 @@
+use core::fmt;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use serde::{Deserialize, Serialize};
+use yew::prelude::*;
+
+use super::{Oneshot, OneshotBridge, OneshotSpawner};
+use crate::utils::get_next_id;
+use crate::worker::WorkerProviderProps;
+use crate::{Bincode, Codec, Reach};
+
+pub(crate) struct OneshotProviderState<T>
+where
+    T: Oneshot + 'static,
+{
+    id: usize,
+    spawn_bridge_fn: Rc<dyn Fn() -> OneshotBridge<T>>,
+    reach: Reach,
+    held_bridge: Rc<RefCell<Option<OneshotBridge<T>>>>,
+}
+
+impl<T> fmt::Debug for OneshotProviderState<T>
+where
+    T: Oneshot,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("OneshotProviderState<_>")
+    }
+}
+
+impl<T> OneshotProviderState<T>
+where
+    T: Oneshot,
+{
+    fn get_held_bridge(&self) -> OneshotBridge<T> {
+        let mut held_bridge = self.held_bridge.borrow_mut();
+
+        match held_bridge.as_mut() {
+            Some(m) => m.fork(),
+            None => {
+                let bridge = (self.spawn_bridge_fn)();
+                *held_bridge = Some(bridge.fork());
+                bridge
+            }
+        }
+    }
+
+    /// Creates a bridge, uses "fork" for public agents.
+    pub fn create_bridge(&self) -> OneshotBridge<T> {
+        match self.reach {
+            Reach::Public => {
+                let held_bridge = self.get_held_bridge();
+                held_bridge.fork()
+            }
+            Reach::Private => (self.spawn_bridge_fn)(),
+        }
+    }
+}
+
+impl<T> Clone for OneshotProviderState<T>
+where
+    T: Oneshot,
+{
+    fn clone(&self) -> Self {
+        Self {
+            id: self.id,
+            spawn_bridge_fn: self.spawn_bridge_fn.clone(),
+            reach: self.reach,
+            held_bridge: self.held_bridge.clone(),
+        }
+    }
+}
+
+impl<T> PartialEq for OneshotProviderState<T>
+where
+    T: Oneshot,
+{
+    fn eq(&self, rhs: &Self) -> bool {
+        self.id == rhs.id
+    }
+}
+
+/// The Oneshot Agent Provider.
+///
+/// This component provides its children access to an oneshot agent.
+#[function_component]
+pub fn OneshotProvider<T, CODEC = Bincode>(props: &WorkerProviderProps) -> Html
```

Comment:
Type parameters should named as `UpperCamelCase`. See [API guidelines](https://rust-lang.github.io/api-guidelines/naming.html)